### PR TITLE
server: Add EngineTypeDefault, resolve it to Pebble if last used

### LINKED ISSUE #42445: cli: make --storage-engine sticky
Make the `--storage-engine` setting sticky so that if you create a store with `--storage-engine=pebble` and then restart Cockroach without specifying the `--storage-engine` flag, the Pebble storage engine will be used. In addition to remove a mild bit of user surprise when restarting Cockroach, this improves the ergonomics for using `debug` commands as there will no longer be a need to specify `--storage-engine` in order to use the same storage engine that was passed to `start`.

### PR DESCRIPTION
**Pull request by itsbilal**
Adds a new engine type called EngineTypeDefault, that resolves
at start time to RocksDB, or Pebble if that was last used
to write to the first store. This effectively makes the
--storage-engine flag sticky for future runs where no
engine is specified. If the flag is passed in again, that
specified engine type takes priority over the one used last.

Fixes #42445 .

Release note (cli change): Add new "default" option for
--storage-engine flag that respects the engine used last.

### CODE REVIEW

**Comment by itsbilal** [comment]
First commit is #45499 , will rebase it away when that gets merged.

**Review by petermattis — COMMENTED** [review]
Approach looks generally good, though I'm worried about how many places you need to add support for `EngineTypeDefault`. Were all of those additions actually needed, or should we be making sure that the engine type is "resolved" early and instead error out if we get to `server.Start` and the engine type is still "default"?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-:-M1ApD2bAYvLKZpcOe24:b-uunvwy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @knz, @petermattis, and @sumeerbhola)

___
*[pkg/cli/debug.go, line 158 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1AiyzH2RaU5Wu63cfJ:-M1AiyzH2RaU5Wu63cfK:bmhnig5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e61b757502b6529577c7a225cfda6992f64a506f/pkg/cli/debug.go#L158)):*
> ```Go
> 		db, err = engine.NewPebble(context.Background(), cfg)
> 
> 	case enginepb.EngineTypeDefault:
> ```

I think this should have the same behavior as `start`. Specifically, we should query `pebble.GetVersion`. The reason for this is so commands like `cockroach debug keys` use the same engine that was previously used. That was actually  my initial motivation behind this functionality.
___
*[pkg/server/server.go, line 1747 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1AozWxB9_WG5FW18A_:-M1AozWxB9_WG5FW18Aa:b-2e86zw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e61b757502b6529577c7a225cfda6992f64a506f/pkg/server/server.go#L1747)):*
> ```Go
> 	case enginepb.EngineTypePebble:
> 		nodeStartCounter += "pebble."
> 	case enginepb.EngineTypeDefault:
> ```

I think we want to know what engine is actually being used at this point. Will we ever get here with `enginepb.EngineTypeDefault` still be a valid case?

**Review by petermattis — COMMENTED** [review]
Also, I think you need to update the help text for the `--storage-engine` flag.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-:-M1ApY9o2q-1kdmFMxoT:b-3d9teb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @knz, @petermattis, and @sumeerbhola)

**Review by itsbilal — COMMENTED** [review]
Thanks for the review! Walked back some of the `Default` checks, and added an error at engine creation time, if the engine type is still left at `EngineTypeDefault`. Test code still continues to treat Default = RocksDB

Also made the help text change.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-:-M1DGAaV59Jq3q9GZGCu:bxv48wv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @petermattis, and @sumeerbhola)

___
*[pkg/cli/debug.go, line 158 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1AiyzH2RaU5Wu63cfJ:-M1DHT1yEyX65jm2h4PC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e61b757502b6529577c7a225cfda6992f64a506f/pkg/cli/debug.go#L158)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think this should have the same behavior as `start`. Specifically, we should query `pebble.GetVersion`. The reason for this is so commands like `cockroach debug keys` use the same engine that was previously used. That was actually  my initial motivation behind this functionality.
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 1747 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1AozWxB9_WG5FW18A_:-M1DHu5lDsJDfC1bv-n3:bgw6vsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e61b757502b6529577c7a225cfda6992f64a506f/pkg/server/server.go#L1747)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think we want to know what engine is actually being used at this point. Will we ever get here with `enginepb.EngineTypeDefault` still be a valid case?
</blockquote></details>

The storage engine value should have been resolved by this point, but just in case it isn't, I added a case for `default` instead of not reporting anything or errorring out, so we can at least find out something's wrong without erroring out.

**Comment by itsbilal** [comment]
Actually, this turned out to be more involved to implement in test code than I expected at first. We'd basically have to take all instances of `MakeConfig` in tests, and append a conditional that goes `if cfg.StorageEngine == Default { cfg.StorageEngine = EngineTypeRocksDB }` (or alternatively we could just set `engine.DefaultStorageEngine` to `EngineTypeRocksDB` if it's `default` right before the `MakeConfig` calls).

I'd now make the case to go back to the previous, defensive solution where we treat `EngineTypeDefault` as akin to `EngineTypeRocksDB` whenever it's encountered, except in start and debug where we make the `GetVersion` check. It might seem like more threading, but it's less threading than the alternative, plus it feels safer.

**Review by petermattis — COMMENTED** [review]
> I'd now make the case to go back to the previous, defensive solution where we treat EngineTypeDefault as akin to EngineTypeRocksDB whenever it's encountered, except in start and debug where we make the GetVersion check. It might seem like more threading, but it's less threading than the alternative, plus it feels safer.

Ack. Not all of my suggestions are good suggestions. I like this proposal. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-:-M1GJ0863YesnmaLocW_:bjahwge)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @knz, @petermattis, and @sumeerbhola)

___
*[pkg/cli/cliflags/flags.go, line 719 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1GIh3G4muCs6Zq4IZA:-M1GIh3G4muCs6Zq4IZB:bfuzcv6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a17f1aa371133a7b5eb936091701ae5329eb7af/pkg/cli/cliflags/flags.go#L719)):*
> ```Go
> 		Description: `
> Storage engine to use for all stores on this cockroach node. Options are rocksdb,
> or pebble.
> ```

Isn't `default` an option now too?
___
*[pkg/cli/cliflags/flags.go, line 721 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1GIj5XA-Aj4qyzRIe2:-M1GIj5XA-Aj4qyzRIe3:bit2ax0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a17f1aa371133a7b5eb936091701ae5329eb7af/pkg/cli/cliflags/flags.go#L721)):*
> ```Go
> or pebble.
> 
> If left unspecified, the storage engine last used to write to the
> ```

I'd rephrase this to mention "default": `If default is specified`. That applies to the other references to "unspecified" as well.

**Review by itsbilal — COMMENTED** [review]
Thanks for the review, this is ready for another look. I've reverted to the `default = rocksdb` idea. Also added a release note.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-:-M1RBuLm1yHXdYDQzW3t:b-t6o2tu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @petermattis, and @sumeerbhola)

___
*[pkg/cli/cliflags/flags.go, line 719 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1GIh3G4muCs6Zq4IZA:-M1RBgNS-NLMqYcYVkY1:b-xvgw26) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a17f1aa371133a7b5eb936091701ae5329eb7af/pkg/cli/cliflags/flags.go#L719)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Isn't `default` an option now too?
</blockquote></details>

Done. Yeah, it is, but I was trying to simplify it to just the unspecified case. I like explicitly mentioning `default` better, so I made the change.
___
*[pkg/cli/cliflags/flags.go, line 721 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1GIj5XA-Aj4qyzRIe2:-M1RBq2DCSl7lxPoaDay:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a17f1aa371133a7b5eb936091701ae5329eb7af/pkg/cli/cliflags/flags.go#L721)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'd rephrase this to mention "default": `If default is specified`. That applies to the other references to "unspecified" as well.
</blockquote></details>

Done.

**Review by petermattis — APPROVED** [review]
LGTM modulo a couple of nits. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-:-M1RNlJ_992hQzn6mZQy:b20745z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @itsbilal, @knz, and @sumeerbhola)

___
*[pkg/cli/debug.go, line 154 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1RN91X6w3r0d3TZly5:-M1RN91X6w3r0d3TZly6:b-d7xoye) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0d59b853b54b7b9c383f6bcaf2c25844dfa4d0c/pkg/cli/debug.go#L154)):*
> ```Go
> 			}
> 		}
> 	}
> ```

This is duplicated with code in `runStart`, and is just past my threshold where I'd want to extract it into a common method. Something like `resolveStorageEngineType(engine.DefaultStorageEngine, dir)`.
___
*[pkg/storage/engine/engine.go, line 560 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1RNYgF67W8cS5GxdVX:-M1RNYgF67W8cS5GxdVY:bcmp86) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0d59b853b54b7b9c383f6bcaf2c25844dfa4d0c/pkg/storage/engine/engine.go#L560)):*
> ```Go
> 	case enginepb.EngineTypeDefault:
> 		fallthrough
> 	case enginepb.EngineTypeRocksDB:
> ```

Nit: you can write this as `case enginepb.EngineTypeDefault, enginepb.EngineTypeDefault:`.
___
*[pkg/storage/engine/in_mem.go, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1RNdm33fsNkIrEu-Az:-M1RNdm33fsNkIrEu-B-:b-vkmjwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0d59b853b54b7b9c383f6bcaf2c25844dfa4d0c/pkg/storage/engine/in_mem.go#L35)):*
> ```Go
> 	case enginepb.EngineTypeDefault:
> 		fallthrough
> 	case enginepb.EngineTypeRocksDB:
> ```

See nit about multi-version case statements. 
___
*[pkg/storage/engine/temp_engine.go, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1RNhF61vqWcbcsVbBZ:-M1RNhF61vqWcbcsVbB_:b-vkmjwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0d59b853b54b7b9c383f6bcaf2c25844dfa4d0c/pkg/storage/engine/temp_engine.go#L42)):*
> ```Go
> 	case enginepb.EngineTypeDefault:
> 		fallthrough
> 	case enginepb.EngineTypeRocksDB:
> ```

See nit about multi-version case statements.

**Review by itsbilal — COMMENTED** [review]
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-:-M1RnKqX-6Z35owazr0x:b-9qdoz3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @petermattis, and @sumeerbhola)

___
*[pkg/cli/debug.go, line 154 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1RN91X6w3r0d3TZly5:-M1RnJ7v0cid0Jl7eTcY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0d59b853b54b7b9c383f6bcaf2c25844dfa4d0c/pkg/cli/debug.go#L154)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is duplicated with code in `runStart`, and is just past my threshold where I'd want to extract it into a common method. Something like `resolveStorageEngineType(engine.DefaultStorageEngine, dir)`.
</blockquote></details>

Done.
___
*[pkg/storage/engine/in_mem.go, line 35 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1RNdm33fsNkIrEu-Az:-M1RiP6NEVflJnlr00E2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0d59b853b54b7b9c383f6bcaf2c25844dfa4d0c/pkg/storage/engine/in_mem.go#L35)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

See nit about multi-version case statements. 
</blockquote></details>

Done.
___
*[pkg/storage/engine/temp_engine.go, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45512#-M1RNhF61vqWcbcsVbBZ:-M1RiPV3DqK1ey1XKuGm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0d59b853b54b7b9c383f6bcaf2c25844dfa4d0c/pkg/storage/engine/temp_engine.go#L42)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

See nit about multi-version case statements. 
</blockquote></details>

Done.

**Review by petermattis — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45512)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @sumeerbhola)

**Comment by itsbilal** [comment]
TFTR!

bors r+
